stl/inc/memory

```diff
@@ -2416,26 +2437,283 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
+        shared_ptr<_Ty> _Old = exchange(_STD move(_Value), _Order);
+    }
+
+    _NODISCARD shared_ptr<_Ty> load() const noexcept {
+        shared_ptr<_Ty> _Result = _Base::load();
+        _Result._Incref();
+        return _Result;
+    }
+
+    _NODISCARD shared_ptr<_Ty> load(const memory_order _Order) const noexcept {
+        shared_ptr<_Ty> _Result = _Base::load(_Order);
+        _Result._Incref();
+        return _Result;
+    }
+
+    operator shared_ptr<_Ty>() const noexcept {
+        return load();
+    }
+
+    _NODISCARD shared_ptr<_Ty> exchange(shared_ptr<_Ty> _Value) noexcept {
+        _Value._Incref();
+        return _Base::exchange(_STD move(_Value));
+    }
+
+    _NODISCARD shared_ptr<_Ty> exchange(shared_ptr<_Ty> _Value, const memory_order _Order) noexcept {
+        _Value._Incref();
+        return _Base::exchange(_STD move(_Value), _Order);
+    }
+
+    bool compare_exchange_weak(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Success,
+        const memory_order _Failure) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Combine_cas_memory_orders(_Success, _Failure));
+    }
+
+    bool compare_exchange_strong(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Success,
+        const memory_order _Failure) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Combine_cas_memory_orders(_Success, _Failure));
+    }
+
+    bool compare_exchange_weak(
+        shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired, const memory_order _Order) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired), _Order);
+    }
+
+    bool compare_exchange_strong(
+        shared_ptr<_Ty>& _Expected, const shared_ptr<_Ty> _Desired, const memory_order _Order) noexcept {
+        _Ref_count_base* _ExpectedRep = _Expected._Rep;
+        const bool _Result            = _Base::compare_exchange_strong(_Expected, _Desired, _Order);
+        if (_Result) {
+            _Desired._Incref();
+            _Expected._Decref();
+        } else {
+            _Expected._Incref();
+            if (_ExpectedRep) {
+                _ExpectedRep->_Decref();
+            }
+        }
+        return _Result;
+    }
+
+    bool compare_exchange_weak(shared_ptr<_Ty>& _Expected, shared_ptr<_Ty> _Desired) noexcept {
+        return compare_exchange_strong(_Expected, _STD move(_Desired));
+    }
+
+    bool compare_exchange_strong(shared_ptr<_Ty>& _Expected, const shared_ptr<_Ty> _Desired) noexcept {
+        _Ref_count_base* _ExpectedRep = _Expected._Rep;
+        const bool _Result            = _Base::compare_exchange_strong(_Expected, _Desired);
+        if (_Result) {
+            _Desired._Incref();
+            _Expected._Decref();
+        } else {
+            _Expected._Incref();
+            if (_ExpectedRep) {
+                _ExpectedRep->_Decref();
+            }
+        }
+        return _Result;
+    }

```

Comment:
Note for other reviewers:

I didn't look at these functions due to not understanding atomics well enough